File: JavaScript/array/MoveZeros.js
/**
 * Time: O(n)
 * Space: O(n)
 * 
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var moveZeroes = function(nums) {
  let count = 0;
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] === 0) {
      count++;
    }
  }
  
  let newNums = [];
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] !== 0) {
      newNums.push(nums[i]);
    }
  }
  
  while (count > 0) {
    newNums.push(0);
    count--;
  }
  
  for (let i = 0; i < newNums.length; i++) {
    nums[i] = newNums[i];
  }
};

/**
 * Time: O(n)
 * Space: O(1)
 * 
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var moveZeroes = function(nums) {
  let lastNoneZeros = 0;
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] !== 0) {
      nums[lastNoneZeros++] = nums[i];
    }
  }
  for (let i = lastNoneZeros; i < nums.length; i++) {
    nums[i] = 0;
  }
};

File: Python/array/MoveZeros.py
# -*- coding:UTF-8 -*-
class Solution:
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        @ Solution:
        - 使用两个指针，index和cur；
        - index指向数组中第一个为0的位置，并且小于cur；
        - cur指向当前正在遍历的值，如果当前值不为0，并且index指向0，二者进行交换；
        """
        if len(nums) > 1:
            index, cur = 0, 0
            while cur < len(nums):
                if nums[cur] != 0 and nums[index] == 0:
                    nums[index] = nums[cur]
                    nums[cur] = 0
                cur += 1
                while nums[index] != 0 and index < cur-1:
                    index += 1
                